### PR TITLE
Add a PrintCommunicator to print progress from the CLI

### PR DESCRIPTION
**Pull request by pklampros**
Prints this for the axial analysis of the barnsbury_extended2_axial.graph:

> Loading graph barnsbury_extended2_axial.graph ok
> Running axial analysis... step: 0/0 record: 1092/7987
> step: 0/0 record: 2166/7987
> step: 0/0 record: 3117/7987
> step: 0/0 record: 4052/7987
> step: 0/0 record: 5123/7987
> step: 0/0 record: 6182/7987
> step: 0/0 record: 7238/7987
> ok
> Writing out result... ok

### CODE REVIEW

File: depthmapXcli/commandlineparser.cpp [hunk 1]
```diff
@@ -25,11 +25,12 @@
 using namespace depthmapX;
 
 void CommandLineParser::printHelp(){
-    std::cout << "Usage: depthmapXcli -m <mode> -f <filename> -o <output file> [-s] [mode options]\n"
+    std::cout << "Usage: depthmapXcli -m <mode> -f <filename> -o <output file> [-s] [-t] [-p] [mode options]\n"
```
Comment (blsemo) [inline depthmapXcli/commandlineparser.cpp:28]: If adding `-t` to the help output, it should be `[-t <times.csv>]`. This option requires a file argument.

File: depthmapXcli/commandlineparser.cpp [hunk 2]
```diff
@@ -108,6 +109,10 @@ void CommandLineParser::parse(size_t argc, char *argv[])
         {
             m_simpleMode = true;
         }
+        else if ( std::strcmp("-p", argv[i]) == 0)
```
Comment (blsemo) [inline depthmapXcli/commandlineparser.cpp:112]: Is printing progress not available on simple mode? The `else if` suggests that...

File: depthmapXcli/runmethods.cpp [hunk 3]
```diff
@@ -53,6 +54,13 @@ namespace dm_runmethods
         return mgraph;
     }
 
+    Communicator* getCommunicator(const CommandLineParser &clp) {
+        if (clp.printProgress()) {
+            return new PrintCommunicator();
```
Comment (blsemo) [inline depthmapXcli/runmethods.cpp:59]: Creating a communicator with `new` that never gets deleted? I know that this most likely get destroyed when the client program shuts down, but this is still a dirty resource leak. Should be assigned to a smart pointer.

**Review by blsemo — CHANGES_REQUESTED** [review]
The idea looks good in principle, but there are a few smallish issues...

File: depthmapXcli/commandlineparser.cpp (hunk 2 above)
Comment (pklampros) [reply depthmapXcli/commandlineparser.cpp:112]: The `else if` examines whether a given argument is `-s` or `-p`, so providing both  as different arguments should run simple mode with progress prints

File: depthmapXcli/runmethods.cpp [hunk 4]
```diff
@@ -53,6 +54,13 @@ namespace dm_runmethods
         return mgraph;
     }
 
+    std::unique_ptr<Communicator> getCommunicator(const CommandLineParser &clp) {
+        if (clp.printProgress()) {
+            return std::unique_ptr<Communicator>(new PrintCommunicator());
```
Comment (blsemo) [inline depthmapXcli/runmethods.cpp:59]: I am not sure if the unique pointer is required here - wouldn't returning a raw pointer wrap it automatically? But probably doesn't make a difference as the return value optimisation will ensure there's just one unique pointer created and moved.

**Review by blsemo — COMMENTED** [review]
Looks ok to me now, but needs some investigation why a few of the regression tests seem to crash now.

**Pull request merged**